zcash_client_sqlite/src/wallet.rs

```diff
@@ -166,35 +267,115 @@ pub(crate) fn memo_repr(memo: Option<&MemoBytes>) -> Option<&[u8]> {
+) -> Result<Option<zip32::AccountId>, SqliteClientError> {
+    conn.query_row_and_then(
+        "SELECT MAX(hd_account_index) FROM accounts WHERE hd_seed_fingerprint = :hd_seed",
+        [seed_id.as_bytes()],
+        |row| {
+            let account_id: Option<u32> = row.get(0)?;
+            account_id
+                .map(zip32::AccountId::try_from)
+                .transpose()
+                .map_err(|_| SqliteClientError::AccountIdOutOfRange)
+        },
+    )
+}
+
+struct AccountSqlValues<'a> {
+    account_type: u32,
+    hd_seed_fingerprint: Option<&'a [u8]>,
+    hd_account_index: Option<u32>,
+    ufvk: Option<String>,
+    uivk: String,
+}
+
+/// Returns (account_type, hd_seed_fingerprint, hd_account_index, ufvk, uivk) for a given account.  
+fn get_sql_values_for_account_parameters<'a, P: consensus::Parameters>(
+    account: &'a Account,
+    params: &P,
+) -> Result<AccountSqlValues<'a>, SqliteClientError> {
+    Ok(match account {
+        Account::Zip32(hdaccount) => AccountSqlValues {
+            account_type: AccountType::Zip32.into(),
+            hd_seed_fingerprint: Some(hdaccount.hd_seed_fingerprint().as_bytes()),
+            hd_account_index: Some(hdaccount.account_index().into()),
+            ufvk: Some(hdaccount.ufvk().encode(params)),
+            uivk: ufvk_to_uivk(hdaccount.ufvk(), params)?,
+        },
+        Account::Imported(ImportedAccount::Full(ufvk)) => AccountSqlValues {
+            account_type: AccountType::Imported.into(),
+            hd_seed_fingerprint: None,
+            hd_account_index: None,
+            ufvk: Some(ufvk.encode(params)),
+            uivk: ufvk_to_uivk(ufvk, params)?,
+        },
+        Account::Imported(ImportedAccount::Incoming(uivk)) => AccountSqlValues {
+            account_type: AccountType::Imported.into(),
+            hd_seed_fingerprint: None,
+            hd_account_index: None,
+            ufvk: None,
+            uivk: uivk.encode(&params.network_type()),
+        },
     })
 }
 
+pub(crate) fn ufvk_to_uivk<P: consensus::Parameters>(
+    ufvk: &UnifiedFullViewingKey,
+    params: &P,
+) -> Result<String, SqliteClientError> {
+    let mut ivks: Vec<Ivk> = Vec::new();
+    if let Some(orchard) = ufvk.orchard() {
+        ivks.push(Ivk::Orchard(orchard.to_ivk(Scope::External).to_bytes()));
+    }
+    if let Some(sapling) = ufvk.sapling() {
+        let dk = sapling.to_bytes();
+        let ivk = sapling.to_ivk(Scope::External).to_repr();
+        let mut divk = [0u8; 64];
+        divk[..32].copy_from_slice(&dk[96..]);
+        divk[32..].copy_from_slice(&ivk[..]);
+        ivks.push(Ivk::Sapling(divk));
+    }
+    if let Some(transparent) = ufvk.transparent() {
+        ivks.push(Ivk::P2pkh(transparent.serialize().try_into().map_err(
+            |_| {
+                SqliteClientError::BadAccountData(
+                    "Unable to serialize transparent IVK.".to_string(),
+                )
+            },
+        )?));
```

Comment:
Bug: This is serializing the transparent FVK, not the external transparent IVK.

Reply:
It's funny you bring this up, because in an earlier iteration I had the code first get the external transparent IVK. But that caused tests to fail when they took the UIVK's transparent component and repeated the external IVK derivation. Maybe I 'fixed' the test in the wrong place. I don't think of transparent keys offering an IVK option, so I just stored the whole public key in there. 
But now that you mention it, I look at ZIP-316 and find this:
> For a Transparent P2PKH FVK, the corresponding Transparent P2PKH IVK is obtained by deriving the child key with non-hardened index 
 as described in [32](https://zips.z.cash/zip-0316#bip-0032-public-to-public).

That seems to verify your expectation. I'll make the change.